Design communication 5 (#115)
SW709E19/Report: report/PartDesign/ChapterCommunication/Communication.tex

\chapter{Communication}\label{cha:communication}
When considering how to design communication between the devices, there are some important attributes that we need to keep in mind.

First off, the sensor device will be making measurements of upwards of 300hz as seen in chapter \ref{cha:prototype}, thus, it is important that we can send data fast enough.
Sending every single measurement separately would be very inefficient, and we should instead collect and send the data as blocks instead.

Another important thing is that the data transfer has to be mostly lossless to be usable for predictive maintenance.
A solution to this would be to utilise persistent storage where possible.

> The memory on the gateway is large enough to store the collected data for quite a while in case it loses connection to the cloud service.

Reviewer: ```suggestion
The memory on the gateway is large enough to store the collected data for some time in case it loses connection to the cloud service.
```

The sensor device can also store a small amount of data in case it loses connection to the gateway, however, its memory is much smaller than the gateways.

power consumption is also something that should be kept in mind since the sensor device will be powered by a battery.

> If we constantly send data we will empty the battery very quickly, thus we need to collect and send data in smaller bursts, and then have the sensor device sleep until the next burst.

Reviewer: ```suggestion
If the device constantly sends data the battery will not last very long, thus it needs to collect and send data in smaller bursts, and then have the sensor device sleep until the next burst.
```

Reviewer: Burst?

> We still need to collect enough data to perform predictive maintenance, so on the other hand, there also should not be too long between the bursts.

Reviewer: But is it a problem if the data might by a little late. The data analysis does not require the data to be realtime.

Author: ```suggestion
We still need to collect enough data to perform predictive maintenance, so on the other hand, there also should not be too long between the data collection.
```

> One option would be to configure the sensor device to only wake up and collect data during the timeframe that the machine is active, for the instances where this information is known.

Reviewer: ```suggestion
One option would be to configure the sensor device to only wake up and collect data during the timeframe that the machine is active, for situations where this information is known.
```

> Both the sensor device and the gateway also need to be able to send data back and forth wirelessly.

Reviewer: ```suggestion
Both the sensor device and the gateway needs to be able to send data back and forth wirelessly.
```

> In some places, the gateway might have a wired connection, but we cannot depend on it.

Reviewer: ```suggestion
In some places, the gateway might have a wired connection, however this might not be applicable in every situation.
```

> We do not know the distance between the sensor device and the gateway, thus we must make sure that whatever we chose, it has enough range to maintain a stable connection.

Reviewer: ```suggestion
The distance between the sensor device and the gateway are not know before hand, thus the devices should have enough range to maintain a stable connection.
```

> However, increased range and data speed come at the cost of increased power consumption.

Reviewer: ```suggestion
However, increased range and data speed comes at the cost of increased power consumption.
```

The sensor device has the option of both Bluetooth low energy and Wi-Fi.

> And while Wi-Fi does use more energy, it is still the better choice for us because of the higher data speed.

Reviewer: ```suggestion
While Wi-Fi does use more energy, it is still a more sustainable choice due to the higher data speed.
```

As section \ref{sec:configurationofdevices} mentioned, it is necessary to be able to set configurations of the sensor devices remotely, whilst they, in turn, send data to the cloud.

> For this, we need a data protocol.

Reviewer: ```suggestion
In order to achieve the above mentioned attributes a suitable communications protocol is needed.
```

> The data protocol must be lightweight since the sensor device has limited memory.

Reviewer: ```suggestion
The communication protocol must be lightweight since the sensor device has limited memory.
```

> It is also important that it is stable and that it doesn't consume too much power.

Reviewer: ```suggestion
It is important that  the protocol is stable and that it consumes little power.
```

There are a lot of options for data protocols, but in the end, we have settled for MQTT/Mosquitto.

> It was chosen because its very lightweight, it doesn't consume a lot of bandwidth, it has a low memory requirement and doesn't use a lot of power.

Reviewer: Should we write that we used MQTT in the design?
Isn't it more a implementation thing?

Author: From the report structure file, it appears that MQTT isn't mentioned until implementation. So just cut that entire last part? Or what do you suggest?

Reviewer: I think it would be best to just cut the last part and add a new finishing part to this section. 
In my head the design should not contain any implementation specific details but should instead be used to describe the intended design, as you did above.
